drivers/nec/display.cr

```diff
@@ -0,0 +1,303 @@
+      self[:input] = Input.from_value(value)
+    when .audio_input?
+      self[:audio] = Audio.from_value(value)
+    when .volume_status?
+      self[:volume] = value
+      self[:audio_mute] = value == 0
+    when .brightness_status?
+      self[:brightness] = value
+    when .contrast_status?
+      self[:contrast] = value
+    when .mute_status?
+      self[:audio_mute] = value == 1
+      self[:volume] = 0 if value == 1
+    when .auto_setup?
+      # auto_setup
+      # nothing needed to do here (we are delaying the next command by 4 seconds)
+    else
+      logger.info { "-- NEC LCD, unknown response: #{data[10..13]}" }
+      logger.info { "-- NEC LCD, full response was: #{data}" }
+    end
+  end
+
+  enum Command
+    VideoInput       = 0x0060
+    AudioInput       = 0x022E
+    VolumeStatus     = 0x0062
+    MuteStatus       = 0x008D
+    PowerOnDelay     = 0x02D8
+    ContrastStatus   = 0x0012
+    BrightnessStatus = 0x0010
+    AutoSetup        = 0x001E
+    PowerQuery       = 0x01D6
+    Save             = 0x0C
+    SetPower         = 0xC203D6
+  end
+
+  private def format_value(value : Int, length : Int = 4) : String
+    length = 2 if value == 0x0C # To deal with Command::Save
+    length = 6 if value == 0xC203D6 # To deal with Command::SetPower
+    value.to_s(16, true).rjust(length, '0')
+  end
+
+  {% for name in Command.constants %}
+    @[Security(Level::Administrator)]
+    def {{name.id.underscore}}(priority : Int32 = 0)
+      send(MsgType::GetParameter.build(Command::{{name.id}}), priority: priority, name: {{name.id.underscore.stringify}})
+    end
+  {% end %}
+
+  private def check_checksum(data : Bytes)
+    checksum = 0x00_u8
+    # Loop through the second to the third last element
+    if data.size >= 2
+      data[1..-3].each do |b|
+        checksum = checksum ^ b
+      end
+      # Check the checksum equals the second last element
+      logger.debug { "Error: checksum should be #{checksum.to_s(16)}" } unless checksum == data[-2]
+      checksum == data[-2]
+    else
+      true
+    end
+  end
+
+  # Types of messages sent to and from the LCD
+  enum MsgType : UInt8
+    Command           = 0x41 # 'A'
+    CommandReply      = 0x42 # 'B'
+    GetParameter      = 0x43 # 'C'
+    GetParameterReply = 0x44 # 'D'
+    SetParameter      = 0x45 # 'E'
+    SetParameterReply = 0x46 # 'F'
+
+    def format_value(value : Int, length : Int = 4) : String
+      length = 2 if value == 0x0C # To deal with Command::Save
+      value.to_s(16, true).rjust(length, '0')
+    end
+
+    def build(command : Nec::Display::Command, data : Int? = nil)
+      message = format_value(command.value)
+
+      message = String.build do |str|
+        str << "0*0"
+        str.write_byte self.value                # Type
+
+        message_length = message.size + 2
+        message_length += 4 if d = data          # If there is data, add 4 to the message length
+        str << format_value(message_length, 2)   # Message length
```

Comment:
Now that some of this formatting is neater to inline the conversion in place of a seperate method call.